features/FEATURE_BLE/ble/pal/PalSecurityManager.h

```diff
@@ -0,0 +1,1017 @@
+     */
+    virtual ble_error_t enable_encryption(
+        connection_handle_t connection,
+        const ltk_t &ltk
+    ) = 0;
+
+    virtual ble_error_t disable_encryption(
+        connection_handle_t connection
+    ) = 0;
+
+    /**
+     * Get the link's key size.
+     *
+     * @param[in] connection connection handle
+     * @param[out] bytesize size of the encryption key in bytes
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t get_encryption_key_size(
+        connection_handle_t connection,
+        uint8_t &bytesize
+    ) = 0;
+
+    /**
+     * Encrypt data with a given key. This uses the facility on the controller to
+     * perform the encryption.
+     *
+     * @param[in] key encryption key
+     * @param[in,out] data data to be encrypted, if successful contains the result
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t encrypt_data(
+        const key_t &key,
+        encryption_block_t &data
+    ) = 0;
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Privacy
+    //
+
+    virtual ble_error_t set_private_address_timeout(
+        uint16_t timeout_in_seconds
+    ) = 0;
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Keys
+    //
+
+    /**
+     * Set the LTK that is to be used for encryption.
+     *
+     * @param[in] connection connection handle
+     * @param[in] ltk long term key
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t set_ltk(
+        connection_handle_t connection,
+        const ltk_t &ltk
+    ) = 0;
+
+    /**
+     * Inform the stack we don't have the LTK.
+     *
+     * @param[in] connection connection handle
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t set_ltk_not_found(
+        connection_handle_t connection
+    ) = 0;
+
+    /**
+     * Set the local IRK.
+     *
+     * @param[in] irk identity resolution key
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t set_irk(
+        const irk_t &irk
+    ) = 0;
+
+    /**
+     * Set the local CSRK.
+     *
+     * @param[in] csrk signing key
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t set_csrk(
+        const csrk_t &csrk
+    ) = 0;
+
+    /**
+     * Generate the Public key. This will also generate the private key.
+     * Public key will be returned as an event handler callback when it's ready.
+     *
+     * @retval BLE_ERROR_NONE On success, else an error code indicating reason for failure
+     */
+    virtual ble_error_t generate_public_key() = 0;
+
```

Comment:
Should be renamed to something like `generate_keypair()` as the key pair is generated.

Reply:
we don't get the private key (and that's the name the spec uses as well)

Reply:
OK, then it's a BLE spec issue 😛 - let's stay consistent with it.